Mathlib/Data/Matroid/DirectSum.lean

```diff
@@ -0,0 +1,269 @@
+
+variable {α : Type*}
+
+lemma Set.union_inters_of_subset_union {A B C : Set α} (hA : A ⊆ B ∪ C) :
+    A ∩ B ∪ A ∩ C = A := by
+  aesop
+
+lemma Set.subset_inter_of_redundant_right {A B C D : Set α} (hAB : A ∪ D ⊆ B) (hBC : A ⊆ C) :
+    A ⊆ B ∩ C :=
+  Set.subset_inter ((Set.subset_union_left A D).trans hAB) hBC
+
+lemma Set.subset_inter_of_redundant_left {A B C D : Set α} (hAB : D ∪ A ⊆ B) (hBC : A ⊆ C) :
+    A ⊆ B ∩ C :=
+  Set.subset_inter ((Set.subset_union_right D A).trans hAB) hBC
+
+lemma Set.union_inter_aux {A B C D : Set α} (hAB : A ⊆ B) : A ∪ C ∩ D ⊆ B ∪ (A ∪ C ∩ D) ∩ D := by
+  simp only [Set.subset_def, Set.mem_inter_iff, Set.mem_union]
+  tauto
+
+lemma Set.inter_union_aux {A B C D : Set α} (hAB : A ⊆ B) : C ∩ D ∪ A ⊆ (C ∩ D ∪ A) ∩ D ∪ B := by
+  simp only [Set.subset_def, Set.mem_inter_iff, Set.mem_union]
+  tauto
+
+lemma Set.union_inter_supset_left_of_disjoint {I₁ I₂ E₁ E₂ : Set α}
+    (hE : E₁ ∩ E₂ = ∅) (hI : I₂ ⊆ E₂) :
+    (I₁ ∪ I₂) ∩ E₁ = E₁ ∩ I₁ := by
+  rw [Set.union_inter_distrib_right]
+  conv => rhs; rw [Set.inter_comm]
+  convert Set.union_empty _
+  simp only [Set.ext_iff, Set.subset_def] at *
+  aesop
+
+lemma Set.union_inter_supset_right_of_disjoint {I₁ I₂ E₁ E₂ : Set α}
+    (hE : E₁ ∩ E₂ = ∅) (hI : I₁ ⊆ E₁) :
+    (I₁ ∪ I₂) ∩ E₂ = E₂ ∩ I₂ := by
+  rw [Set.union_comm]
+  rw [Set.inter_comm] at hE
+  exact Set.union_inter_supset_left_of_disjoint hE hI
+
+lemma Set.subset_iff_subsets_of_disjoint {A B E₁ E₂ : Set α} (hA : A ⊆ E₁ ∪ E₂) :
+    A ⊆ B ↔ A ∩ E₁ ⊆ B ∩ E₁ ∧ A ∩ E₂ ⊆ B ∩ E₂ := by
+  constructor
+  · simp only [Set.subset_def] at *
+    aesop
+  · intro ⟨hABE₁, hABE₂⟩ x hx
+    simp only [Set.ext_iff, Set.subset_def] at *
+    cases hA x hx with
+    | inl hE₁ => exact (hABE₁ x ⟨hx, hE₁⟩).left
+    | inr hE₂ => exact (hABE₂ x ⟨hx, hE₂⟩).left
+
+lemma Set.between_inter {I T X : Set α} (hIT : I ⊆ T) (hTX : T ⊆ X) (E : Set α) :
+    I ∩ E ⊆ T ∩ E ∧ T ∩ E ⊆ X ∩ E :=
+  ⟨Set.inter_subset_inter_left E hIT, Set.inter_subset_inter_left E hTX⟩
+
+lemma Set.union_inter_eq_fst {S₁ S₂ E₁ E₂ X₁ X₂ : Set α}
+    (hE : E₁ ∩ E₂ = ∅) (hX₁ : S₁ ⊆ X₁ ∩ E₁) (hX₂ : S₂ ⊆ X₂ ∩ E₂) :
+    (S₁ ∪ S₂) ∩ E₁ = S₁ := by
+  rw [Set.union_inter_distrib_right]
+  rw [Set.subset_inter_iff] at hX₁ hX₂
+  convert Set.union_empty S₁
+  · symm
+    rw [Set.left_eq_inter]
+    exact hX₁.right
+  · rw [← Set.subset_empty_iff] at hE ⊢
+    apply (Set.inter_subset_inter_left E₁ hX₂.right).trans
+    rwa [Set.inter_comm] at hE
+
+lemma Set.union_inter_eq_snd {S₁ S₂ E₁ E₂ X₁ X₂ : Set α}
+    (hE : E₁ ∩ E₂ = ∅) (hX₁ : S₁ ⊆ X₁ ∩ E₁) (hX₂ : S₂ ⊆ X₂ ∩ E₂) :
+    (S₁ ∪ S₂) ∩ E₂ = S₂ := by
+  rw [Set.union_comm]
+  rw [Set.inter_comm] at hE
+  exact Set.union_inter_eq_fst hE hX₂ hX₁
+
+set_option linter.unusedVariables false in
+/-- Direct sum of matroids as a set operation. -/
+@[nolint unusedArguments]
+def indepDirectSum {M₁ M₂ : IndepMatroid α} (hME : M₁.E ∩ M₂.E = ∅) (I : Set α) : Prop :=
+  ∃ I₁ I₂ : Set α, I₁ ∪ I₂ = I ∧ M₁.Indep I₁ ∧ M₂.Indep I₂
```

Comment:
```suggestion
I ⊆ M₁.E ∪ M₂.E ∧ M₁.Indep (I ∩ M₁.E) ∧ M₂.Indep (I ∩ M₂.E)
```
could this be easier to work with?

Reply:
I think so. Would it be good as a definition tho?

Reply:
Once you have a lemma showing equivalence, it isn't particularly important what the definition is.

Reply:
It is our lemma `indepDirectSum_iff`.

Reply:
The definition suggested by @apnelson1 is almost certainly easier to work with.

Reply:
Yes, it is. However, either way, we provide a translation between these two definitions. The question is which one should be the official definition. The original one seems to be more recognizable for human readers. Do you agree?